rainier-core/src/main/scala/com/stripe/rainier/compute/Context.scala

```diff
@@ -1,10 +1,71 @@
 package com.stripe.rainier.compute
 
-case class Context(density: Real) {
+case class Context(base: Real, batched: Target) {
   val compiler: Compiler = IRCompiler(200, 100)
-  val variables: List[Variable] = RealOps.variables(density).toList
-  lazy val gradient: List[Real] = Gradient.derive(variables, density).toList
 
-  def compileDensity: Array[Double] => Double =
-    compiler.compile(variables, density)
+  val variables: List[Variable] = {
+    val allVariables =
+      RealOps.variables(base) ++
+        RealOps.variables(batched.real) --
+        batched.placeholders.keySet
+    allVariables.toList.sortBy(_.param.sym.id)
+  }
+
+  //convenience method for simple cases like Walkers
+  def compileDensity: Array[Double] => Double = {
+    val baseCF = compiler.compile(variables, base)
+    val batchCF = compileBatched
+
+    val result = { input: Array[Double] =>
+      baseCF(input) + batchCF(input)
+    }
+    result
+  }
+
+  //not terribly efficient
+  private def compileBatched: Array[Double] => Double = {
+    if (batched.placeholders.isEmpty) {
+      compiler.compile(variables, base)
+    } else {
+      val placeholders = batched.placeholders.keys.toList
+      val cf = compiler.compile(variables ++ placeholders, batched.real)
+
+      val nBatches = batched.placeholders.head._2.size
```

Comment:
If `nRows` were a val on Target a la `Target(real, placeholders) {  def nRows = placeholders.head._2.size }` then this line could be `val nBatches = batched.nRows` which would be another indication that batching rearranges a Target to have `nBatches` rows.

Reply:
Good idea, thanks.